src/bucket/BucketManagerImpl.cpp

```diff
@@ -907,90 +922,135 @@ BucketManagerImpl::maybeSetIndex(std::shared_ptr<Bucket> b,
 }
 
 void
-BucketManagerImpl::scanForEviction(AbstractLedgerTxn& ltx, uint32_t ledgerSeq)
+BucketManagerImpl::scanForEvictionLegacySQL(AbstractLedgerTxn& ltx,
+                                            uint32_t ledgerSeq)
 {
     ZoneScoped;
-    if (protocolVersionStartsFrom(ltx.getHeader().ledgerVersion,
-                                  SOROBAN_PROTOCOL_VERSION))
-    {
-        mBucketList->scanForEviction(mApp, ltx, ledgerSeq,
-                                     mBucketListEvictionCounters);
-    }
+    releaseAssert(protocolVersionStartsFrom(ltx.getHeader().ledgerVersion,
+                                            SOROBAN_PROTOCOL_VERSION));
+    mBucketList->scanForEvictionLegacySQL(
+        mApp, ltx, ledgerSeq, mBucketListEvictionCounters, mEvictionStatistics);
 }
 
-medida::Timer&
-BucketManagerImpl::recordBulkLoadMetrics(std::string const& label,
-                                         size_t numEntries) const
+void
+BucketManagerImpl::startBackgroundEvictionScan(uint32_t ledgerSeq)
+{
+    releaseAssert(mApp.getConfig().isUsingBucketListDB());
+    releaseAssert(!mEvictionFuture.valid());
+
+    auto searchableBL = getSearchableBucketListSnapshot();
+    auto const& cfg = mApp.getLedgerManager().getSorobanNetworkConfig();
+    auto const& sas = cfg.stateArchivalSettings();
+
+    using task_t = std::packaged_task<EvictionResult()>;
+    auto task = std::make_shared<task_t>(
+        [bl = move(searchableBL), iter = cfg.evictionIterator(), ledgerSeq, sas,
+         &counters = mBucketListEvictionCounters,
+         &stats = mEvictionStatistics] {
+            return bl->scanForEviction(ledgerSeq, counters, iter, stats, sas);
+        });
+
+    mEvictionFuture = task->get_future();
+    mApp.postOnBackgroundThread(bind(&task_t::operator(), task),
+                                "SearchableBucketListSnapshot: eviction scan");
+}
+
+void
+BucketManagerImpl::resolveBackgroundEvictionScan(
+    AbstractLedgerTxn& ltx, uint32_t ledgerSeq,
+    LedgerKeySet const& modifiedKeys)
 {
-    if (numEntries != 0)
+    ZoneScoped;
+
+    if (!mEvictionFuture.valid())
```

Comment:
My understanding is [std::future::valid()](https://en.cppreference.com/w/cpp/thread/future/valid) returns false if the future has been default initialized or has had a call to `get`. It returns true if the future has been moved from (i.e. `mEvictionFuture = std::packaged_task<EvictionResult()>::get_future()`) but has not yet been called by `get`. Essentially, `valid` returns true if there is ready or pending work, false if the work is done and has been retrieved, or if no work has been set yet.

This particular state is possible on startup, when we're closing our first ledger and haven't had a chance to call `startBackgroundEvictionScan` as part of `closeLedger` yet.

Reply:
So is this only the case for tests then where we close v20 genesis ledger? In practice, genesis is always protocol 0, so it shouldn't need to do any eviction at all, right?

Reply:
No, by "closing our first ledger", I don't mean the genesis ledger, but the first ledger that a node closes since that node's startup. Like during catchup after assuming state, we'd hit this path on the first call to `ledgerClose`, since the assume state path doesn't kick off the background eviction scan.

Reply:
> Like during catchup after assuming state, we'd hit this path on the first call to ledgerClose, since the assume state path doesn't kick off the background eviction scan.

why not? it should, right? otherwise, we'd stall on that first ledger and have to block until the scan is done. Seems like we should kick off the eviction scan in `LedgerManagerImpl::setLastClosedLedger` which is called by catchup